File: android/app/src/main/java/com/mocapsync/app/core/AppLog.kt
package com.mocapsync.app.core

import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.text.SimpleDateFormat
import java.util.ArrayDeque
import java.util.Date
import java.util.Locale

/**
 * 앱 전역 로그 버퍼.
 *
 * 왜 필요한가:
 * 이 프로젝트는 개발자가 실기기 테스트를 직접 할 수 없는 구조입니다.
 * (코드 작성 -> CI 빌드 -> 사용자가 설치/실행 -> 로그 첨부 -> 수정)
 * 그래서 logcat 이 아니라 "앱 안에서 읽고 파일로 뽑을 수 있는 로그"가 1일차부터 필요합니다.
 *
 * 스레드 안전:
 *   write() 는 synchronized 로 보호됩니다. SNTP 스레드/카메라 콜백에서 호출해도 안전합니다.
 *
 * 성능:
 *   로그 한 줄마다 List 를 복사하지 않습니다. [revision] 만 올리고,
 *   UI 가 recomposition 시점에 [snapshot] 을 한 번 호출합니다.
 *   (3단계에서 프레임마다 로그를 남길 때를 대비한 구조)
 */
object AppLog {

    private const val LOGCAT_TAG = "MocapSync"
    private const val MAX_LINES = 4000

    private val timeFmt = SimpleDateFormat("HH:mm:ss.SSS", Locale.US)
    private val buffer = ArrayDeque<String>()

    private val _revision = MutableStateFlow(0)

    /** 로그가 바뀔 때마다 증가. Compose 에서 이 값을 관찰하고 [snapshot] 을 읽으세요. */
    val revision: StateFlow<Int> = _revision.asStateFlow()

    fun d(tag: String, msg: String) = write('D', tag, msg, null)
    fun i(tag: String, msg: String) = write('I', tag, msg, null)
    fun w(tag: String, msg: String) = write('W', tag, msg, null)
    fun e(tag: String, msg: String, t: Throwable? = null) = write('E', tag, msg, t)

    /** 사용자가 "지금 이 순간"을 표시하고 싶을 때 (예: 스톱워치 촬영 시작 직전) */
    fun mark(msg: String) = write('M', "MARK", "──────── $msg ────────", null)

    private fun write(level: Char, tag: String, msg: String, t: Throwable?) {
        val now = Date()
        val head = "${timeFmt.format(now)} $level/$tag: $msg"
        val line = if (t == null) head else head + "\n" + Log.getStackTraceString(t)

        synchronized(buffer) {
            while (buffer.size >= MAX_LINES) buffer.pollFirst()
            buffer.addLast(line)
        }
        _revision.value = _revision.value + 1

        // logcat 에도 남깁니다 (USB 디버깅이 가능한 상황에서는 이게 더 편함)
        when (level) {
            'E' -> Log.e(LOGCAT_TAG, "$tag: $msg", t)
            'W' -> Log.w(LOGCAT_TAG, "$tag: $msg")
            else -> Log.i(LOGCAT_TAG, "$tag: $msg")
        }
    }

    fun snapshot(): List<String> = synchronized(buffer) { buffer.toList() }

    fun size(): Int = synchronized(buffer) { buffer.size }

    fun clear() {
        synchronized(buffer) { buffer.clear() }
        _revision.value = _revision.value + 1
    }
}

File: ios/App/Core/AppLog.swift
import Foundation
import os

/// 앱 내 로그 버퍼.
///
/// 왜 필요한가:
/// 개발자는 실기기를 만질 수 없습니다. Xcode 콘솔도 볼 수 없습니다.
/// (코드 작성 -> CI 빌드 -> 사용자가 설치/실행 -> 로그 첨부 -> 수정)
/// 그래서 Console.app 이 아니라 **앱 안에서 읽고 파일로 뽑을 수 있는 로그**가
/// 1일차부터 필요합니다. 안드로이드 버전의 AppLog 와 같은 역할입니다.
final class AppLog: ObservableObject {

    static let shared = AppLog()

    private static let maxLines = 4000
    private let logger = Logger(subsystem: "com.mocapsync.ios", category: "app")
    private let lock = NSLock()
    private var buffer: [String] = []

    /// UI 가 관찰합니다. 메인 스레드에서만 갱신합니다.
    @Published private(set) var revision: Int = 0

    private lazy var timeFmt: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "HH:mm:ss.SSS"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    private init() {}

    func d(_ tag: String, _ msg: String) { write("D", tag, msg) }
    func i(_ tag: String, _ msg: String) { write("I", tag, msg) }
    func w(_ tag: String, _ msg: String) { write("W", tag, msg) }
    func e(_ tag: String, _ msg: String) { write("E", tag, msg) }

    /// 사용자가 "지금 이 순간"을 표시하고 싶을 때 (예: 스톱워치 촬영 직전)
    func mark(_ msg: String) { write("M", "MARK", "──────── \(msg) ────────") }

    private func write(_ level: String, _ tag: String, _ msg: String) {
        let line = "\(timeFmt.string(from: Date())) \(level)/\(tag): \(msg)"

        lock.lock()
        buffer.append(line)
        if buffer.count > Self.maxLines {
            buffer.removeFirst(buffer.count - Self.maxLines)
        }
        lock.unlock()

        // os_log 에도 남깁니다 (USB 로 Console.app 을 볼 수 있는 상황에서는 이게 편함)
        switch level {
        case "E": logger.error("\(tag, privacy: .public): \(msg, privacy: .public)")
        case "W": logger.warning("\(tag, privacy: .public): \(msg, privacy: .public)")
        default: logger.info("\(tag, privacy: .public): \(msg, privacy: .public)")
        }

        // @Published 는 메인 스레드에서만 건드려야 합니다.
        // 카메라 콜백이나 네트워크 스레드에서 호출될 수 있으므로 반드시 디스패치합니다.
        if Thread.isMainThread {
            revision &+= 1
        } else {
            DispatchQueue.main.async { [weak self] in
                self?.revision &+= 1
            }
        }
    }

    func snapshot() -> [String] {
        lock.lock()
        defer { lock.unlock() }
        return buffer
    }

    var count: Int {
        lock.lock()
        defer { lock.unlock() }
        return buffer.count
    }

    func clear() {
        lock.lock()
        buffer.removeAll()
        lock.unlock()
        if Thread.isMainThread {
            revision &+= 1
        } else {
            DispatchQueue.main.async { [weak self] in self?.revision &+= 1 }
        }
    }
}
